Add generic `Aggregator` processor

danielhuppmann opened the pull request:
This PR is in relation to a request by @jkikstra for a CEDS-emissions-processor that renames and aggregates emissions-variables from IAM results to match the CEDS schema - see the related PR at https://github.com/IAMconsortium/common-definitions/pull/255

The idea is to have a mapping like
```yaml
- dimension: variable
- aggregate:
  - Target Code:
      - Variable A
      - Variable B
```

that can then be called like 
```python
ceds_mapping = AggregationMapping.from_yaml("ceds-mapping/ceds-variable-mapping.yaml")
ceds_aggregator = Aggregator(mapping=ceds_mapping)

ceds_aggregator.apply(df)
```
to apply the aggregation-mapping onto an IamDataFrame `df`.

The method uses the pyam `rename()` method because that actually works (with summation) on all dimensions. This could be extended later for different aggregation-methods, etc.

This could be useful for @lisahligono as part of the scenario-transfer-routine, where the rename-mapping for regions is implemented as an external yaml file and then just called like shown above as part of the processing-and-transfer.

danielhuppmann commented:
Yes, I will still add tests but wanted to hear any comments about feature-design first...

phackstock commented:
As per bilateral discussion, I'd say this is a good way forward.
I'd include the dimension attribute in the yaml file.

dc-almeida commented:
Makes sense to me!

danielhuppmann commented:
Ready for review @phackstock, I have a few questions from your pedantic wisdom:
I defined an attribue `codes` for the `validate_with_definition()` method, but it seems I cannot use `codes` in the field-validators (even with after-mode)?

Apart from that, I'm not sure about a good way to include the file name in a validation-warning - but maybe we can also leave that for a future clean-up and standardization effort across the package, we are probably not 100% consistent. There are also a few overlaps with the region-processing feature, but I wanted to focus on this PR for now.

I will also add a "rename"-config alternative to work with what @lisahligono implemented in https://github.com/IAMconsortium/common-definitions/pull/272.

I will add documentation in a follow-up PR once #470 is merged.

phackstock commented on nomenclature/processor/generic.py:
```diff
@@ -0,0 +1,153 @@
+    def validate_with_definition(self, dsd: DataStructureDefinition) -> None:
+        error = None
+        # check for codes that are not defined in the codelists
+        codelist = getattr(dsd, self.dimension, None)
```
I'd let the error occur right here if the dimension is not found:
```suggestion
        codelist = getattr(dsd, self.dimension)
```

danielhuppmann replied:
I like the current implementation better because this way, I can show the name of the offending file directly in the error message, making it easier to debug in repositories that have many files.

phackstock replied:
Fair point, but in this case the error does not really have anything to do with the aggregator processor but rather with the missing dimension in the DataStructueDefinition.

phackstock commented on nomenclature/processor/generic.py:
```diff
@@ -0,0 +1,153 @@
+        return _codes
+
+    def validate_with_definition(self, dsd: DataStructureDefinition) -> None:
+        error = None
```
If you raise the error below directly, you don't need this
```suggestion
```

danielhuppmann replied:
As above.

phackstock commented on nomenclature/processor/generic.py:
```diff
@@ -0,0 +1,153 @@
+        if codelist is None:
+            error = f"Dimension '{self.dimension}' not found in DataStructureDefinition"
+        elif invalid := codelist.validate_items(self.codes):
+            error = (
```
```suggestion
            raise ValueError(f"The following {self.dimension}s are not defined in the "
                "DataStructureDefinition:\n - " + "\n - ".join(invalid) + "\nin " + str(self.file) + "")

```

danielhuppmann replied:
As above.

phackstock commented on nomenclature/processor/generic.py:
```diff
@@ -0,0 +1,153 @@
+def _validate_items(items, info, _type):
+    duplicates = [item for item, count in Counter(items).items() if count > 1]
+    if duplicates:
+        raise PydanticCustomError(
```
Even though I introduced them, I'd actually say we should move away from using custom errors. The way that I see it, they don't really provide any additional benefit and just make the code harder to read. 
We don't catch specific errors at any higher level and even if we did, they get wrapped in some pydantic error anyway

danielhuppmann replied:
Well, one benefit of the custom errors is that the code in the actual files is shorter, and the concatenation of a readable string with filename, etc. is moved elsewhere.

I agree that the error messages should be standardized and reused, but this PR is not the right place for that discussion.

phackstock commented on nomenclature/processor/generic.py:
```diff
@@ -0,0 +1,153 @@
+here = Path(__file__).parent.absolute()
+
+
+class AggregationItem(BaseModel):
```
I'm not sure this class is actually needed. 
I think you could just use: `mapping: dict[str, list[str]]`. This way you also don't need the `rename_mapping` property below since you can use `mapping` directly.

danielhuppmann replied:
I'd prefer follow the current implementation in the RegionProcessor, and then see how we can streamline or refactor to use common utils methods/classes...

phackstock commented on nomenclature/processor/generic.py:
```diff
@@ -0,0 +1,153 @@
+            raise ValueError(error + "\nin " + str(self.file) + "")
+
+    @classmethod
+    def from_file(cls, file: Path | str):
```
If you rename the `mapping` attribute to `aggregate` and use `dict[str, list[str]]` in favor of `AggregationItem`, I think you should be able to remove this function almost entirely.

phackstock replied:
Yes, exactly

phackstock reviewed (commented):
I would say that there's potential to simplify and shorten the code.
I have left some comments where exactly to do so in line.

phackstock commented:
> I defined an attribue codes for the validate_with_definition() method, but it seems I cannot use codes in the field-validators (even with after-mode)?

That might not be possible, I'll play around with the code a bit more to see.

phackstock commented on nomenclature/processor/generic.py:
Before I forget it, I'd suggest to rename to `aggregator.py`

danielhuppmann replied:
Thanks, done!

danielhuppmann commented on nomenclature/processor/generic.py:
```diff
@@ -0,0 +1,153 @@
+                f"The following {self.dimension}s are not defined in the "
+                "DataStructureDefinition:\n - " + "\n - ".join(invalid)
+            )
+        if error:
```
As above.

danielhuppmann commented on nomenclature/processor/generic.py:
```diff
@@ -0,0 +1,153 @@
+                "DataStructureDefinition:\n - " + "\n - ".join(invalid)
+            )
+        if error:
+            raise ValueError(error + "\nin " + str(self.file) + "")
```
As above.

danielhuppmann commented on nomenclature/processor/generic.py:
```diff
@@ -0,0 +1,153 @@
+    """Aggregation or renaming of an IamDataFrame on a `dimension`"""
+    file: FilePath
+    dimension: str
+    mapping: list[AggregationItem]
```
Good suggestion, thanks.

danielhuppmann commented:
Thanks for your comments @phackstock!

I renamed file names and modules to **aggregator** and implemented some suggestions, but I'd push larger refactoring like using standardized error messages to a follow-up PR.

phackstock reviewed (approved):
Thanks for incorporating the changes @danielhuppmann. 
Agreed that we should push for a larger refactor and agreed that this PR is not the time and place for that.
So good to be merged from my side.
